rfd/0216-tbot-resiliency.md

```diff
@@ -0,0 +1,214 @@
+method. In effect, it makes creating an API client a *blocking* operation, and
+makes it impossible to create a client without having a working connection to
+the auth server; which differs from gRPC's standard behavior of establishing
+connections on-demand - this difference will become relevant when evaluating our
+implementation options below.
+
+### UX Question: How should we handle connectivity issues on-startup?
+
+`tbot`'s current behavior of refusing to start if the auth server is unavailable
+makes it impossible to implement any kind of graceful degradation.
+
+However, if we proceed without a working connection, we cannot as effectively
+surface configuration issues to users.
+
+So the question becomes: what is the right balance? If a proxy address
+is syntactically valid but plainly wrong, is it best to exit loudly or retry
+indefinitely? Is the answer different for one-shot and daemon mode?
+
+### UX / Implementation Question: Can we remove the path selection logic?
+
+It would also greatly simplify things if we could take the `proxy_server` and
+`auth_server` configuration options at face value, rather than trying both
+connection methods regardless.
+
+If we knew up-front whether the given address is an auth server or a proxy, we
+wouldn't need to test the connection to find out. Instead, we could just create
+the client and rely on gRPC to manage retrying failed connections asynchronously.
+
+Theoretically, we might be able to achieve this and keep the path selection
+logic by implementing a custom gRPC dialer, resolver, etc. (see below) but this
+is a lot of complexity with dubious benefit.
+
+### UX Option 1: Just keep running
+
+Perhaps the most obvious option if connection fails on-startup is to just keep
+running, logging errors, and retrying the connection in the background.
+
+The `tbot` process itself would appear healthy, so as an operator you'd need to
+look at the logs to find out something is wrong. Or ideally, we'd have a useful
+health-check endpoint scraped regularly by your monitoring solution of choice.
+
+Another benefit of having the process actually exit is that it might reset any
+state that is causing the connection to fail, which we'd lose with this option.
+
+### UX Option 1a: Keep running... for a while
+
+An extension of Option 1 would be to keep running but "time out" and terminate
+the process if a connection to the auth server could not be established within a
+chosen timeframe.
+
+This feels like it would be a bit of a ticking timebomb, though.
+
+### UX Option 1b: Expose a way to explicitly test the connection
+
+Many tools provide a way to "dry run" or test your configuration ahead of time,
+for example: `nginx -t`. Perhaps we could add a similar option to the `tbot` CLI?
+Or perhaps one-shot mode already achieves this?
+
+### UX Option 2: Test the connection on first run only
+
+Given `tbot` already stores state (e.g. identity certificates) for use between
```

Comment:
🤔  What if rather than writing some new state, we used the presence of the existing state to determine whether to handle the initial connect/join fatally or resiliently.